File: solutions/erlang/nucleotide-count/1/src/nucleotide_count.erl
-module(nucleotide_count).

-export([count/2, nucleotide_counts/1]).


count(Strand, Nucleotide) -> count(Strand, Nucleotide, 0).

nucleotide_counts(Strand) -> nucleotide_counts(Strand, 0, 0, 0, 0).

%%%%

count([], _, N) -> N;
count([Head | Tail], [Head], N) -> count(Tail, [Head], N + 1);
count([Head | Tail], Nucleotide, N) 
  when Head == $A orelse Head == $C orelse Head == $G orelse Head == $T -> 
    count(Tail, Nucleotide, N).

nucleotide_counts([], A, C, G, T) -> [{"A", A}, {"C", C}, {"G", G}, {"T", T}];
nucleotide_counts([$A | Tail], A, C, G, T) -> nucleotide_counts(Tail, A+1, C, G, T);
nucleotide_counts([$C | Tail], A, C, G, T) -> nucleotide_counts(Tail, A, C+1, G, T);
nucleotide_counts([$G | Tail], A, C, G, T) -> nucleotide_counts(Tail, A, C, G+1, T);
nucleotide_counts([$T | Tail], A, C, G, T) -> nucleotide_counts(Tail, A, C, G, T+1).

File: solutions/cpp/nucleotide-count/1/nucleotide_count.cpp
#include "nucleotide_count.h"
#include <algorithm>
#include <stdexcept>

namespace dna {

counter::counter(const std::string& strain) : strain{strain} {
   if (!all_of(strain.begin(), strain.end(), is_nucleotide))
        throw std::invalid_argument{""};
}

int counter::count(char nucleotide) const {
    if (!is_nucleotide(nucleotide)) throw std::invalid_argument{""};
    return std::count(strain.begin(), strain.end(), nucleotide); 
}

NucleotideStats counter::nucleotide_counts() const {
    NucleotideStats stats { {'A', 0}, {'T', 0}, {'C', 0}, {'G', 0} };
    for (char nucleotide : strain) ++stats[nucleotide];
    return stats;
}

bool counter::is_nucleotide(char ch) {
    return ch == 'A' || ch == 'T' || ch == 'C' || ch == 'G';
}
 
}

